### DISCUSSION TITLE
RFC: `pygen` - Would a Python-centric, lightweight "spiritual fork" of `projen` make sense?

### DISCUSSION THREAD

**Discussion by phitoduck** [projen/projen#3858] [General]
> **TL;DR** The DevEx for `projen` in non-TS languages isn't as good. What would we lose if we had slimmer, Python-centric tool? Or could/should we modify `projen`?

I think `projen` tries to be 2 things:

1. a `Component` engine for managing files with support for escape hatching
2. a opinionated set of "higher level components", e.g. `GitHubProject`

In my experience running an ML platform team, we didn't end up using any of the existing `projen` components aka [2].

What if a Python-based tool existed that only implemented [1]?

E.g.
- `ObjectFile` (and its derivatives) and JSON patch escape hatching
- `TemplatizedFile` (tamperproof file with Jinja-powered escape hatch system, could even be a standalone lib/plugin)
- `SampleFile`
`.projen/files.json` state management to delete stale projen-managed files
Everything else would be up to dev teams to remake per their own opinions, e.g. GitHub Actions, task runners, Python packages, pyproject.toml, etc.

---

## Pros/Cons of forking

**Pros:**
- lower learning curve to modify components for your team (since they're in your language)
- better step debugging experience
- slightly faster synth time 🤣
- private attributes no longer inaccessible due to JSII (e.g. enables the easier creation of higher order components on top of object files)

**Cons:**
- lose access to libraries of reusable components maintained by others. I think this is the key issue. The merit of creating a spiritual fork depends the importance of this.
- advances in `projen`'s `Component` abstraction would have to be manually adapted to the spiritual forks (is this common, though?)
- potentially harder to get buy in, e.g. I can't say "Serious people use projen. See? It's maintained by AWS."

**Hybrid pro-cons**
- Flexibility. 
  - Good: "Cookiecutter devs" are used to building templates from the ground up, w/o depending of 3rd party components.
  - Bad: We miss out on a "one right way" to do things.

---

## 1 - Background from my use cases

I do ML Platform engineering. I love `projen` and the whole #TemplatesAreEvil philosophy.

### 1.1 - `projen` is a great fit for MLOps/DataSci

Data Scientists tend not to be strong software engineers.  ML Platform engineering is all about reducing the cognitive load for DS so that they can self-serve with good engineering practices... and still have time to do their core job.

There are a lot of "MLOps templates" out there, and they all suffer from the "template drift" problem that projen solves. 

**Anecdote:** I worked on a team that was crippled in its ability to iterate on MLOps tools. It was a ~1-month endeavor to introduce platform-level changes to all of our preexisting repos. 

Projen saved us! (20+ DE's, DSs, and MLEs all using projen in Python)

### 2.1 - Need to install NodeJS in CI and locally

Mildly annoying. JSII makes rendering slower than technically needs to be 🤷 

### 2.2 - `projen new` doesn't support projects contributed in Python

We used `cookiecutter` to bootstrap our initial `.projenrc.py` and minimal `Makefile` that installed the Node/Python dependencies to run it.

### 2.3 - Inferior escape hatching support

We wanted all DE, MLOps, and MLE folks to be able/willing to contribute to our components. So **we authored all our internal components in Python** and vended them as a pip package.

**The JSII doesn't expose private properties in the Python bindings.** 

E.g. The actual contents of the `ObjectFile` are not reachable from Python.

**Anecdote:** We needed to patch arrays, e.g. in a GitHub Actions `YamlFile`, there is `workflow.someJob.steps[]`. We could never know for sure which index to modify. And we had no ability to inspect the content to find out.

This became reason the DS went from loving to resenting `projen`. ("If we use your tool, we're locked in to a setup that may not make sense for our project!")

### 2.4 - Not great support for "multi-package mono repos" 

> **TL;DR** - our project setup was so different from `PythonProject`, the only projen-specific features we used were escape hatching and the file management system in `.projen/`.

Our projects generally looked like this:

```txt
my-repo/
  .github/workflows/
    ci.yaml <-- lint/format/test each of the below packages in parallel
  infra/       <-- Python pkg: AWS CDK, CDKtf, or Pulumi
  train/       <-- Python pkg: for training a model
  serve/      <-- Python pkg: defines a REST API to serve a model, depends on `train`
    Dockerfile <-- different versions for tools like FastAPI, BentoML, SageMaker endpoints, etc. (user choice)
  demo/     <-- Python pkg: a UI, e.g. a streamlit/gradio app to play with a served model
    Dockerfile <-- optimized for streamlit or gradio (user choice)
    Makefile    <-- `make install` (`pip install -e .[dev]`); `make lint` (`pre-commit run --all-files`); `make test`, `make synth`, ...
  .pre-commit-config.yaml (ignores any projen-managed files)
```

Projen has a `PythonProject` component, but I found this created a `.projen` folder in each package.

- was a lot to look at 😅
- not ideal, because a "Project" might have multiple packages. Some global files like 
  - `.pre-commit-config.yaml` needed to reflect information about all packages, e.g. "have `black` to ignore any files generated with projen across all Python packages"
  - `ci.yaml` needed to include a `"lint/format/test"`, section for each `PythonPackage`. Convenient to have these in one workflow rather than multiple.'
  - we landed on 2 core abstractions: `class Repo(Project)` and `class PythonPackage(Component)`

### 2.5 No standard solution for escape hatching non-`ObjectFile`s

Had to add the following components. Python users are familiar with the Jinja templating engine. For non-`ObjectFile`s like `Makefile`, `Dockerfile`, 
  - `TemplatizedFile(template_rel_fpath: str, out_rel_fpath: str, values_to_substitute: dict[str, Any])`
    - Generates a Jinja template `SampleFile` at `template_rel_fpath` with blocks that the project owner can override
  - `FastAPIDockerfile`, `StreamlitDockerfile`, `Makefile` each wrapped `TemplatizedFile` and exposed, e.g. a `Dockerfile.jinja` template for escape hatching.

**Comment by mrgrain** [comment]
> 2.1 - Need to install NodeJS in CI and locally
>
> Mildly annoying. JSII makes rendering slower than technically needs to be 🤷

I'm painfully aware but cannot offer a solution for it. Given the current situation, I think the options are to either build polyglot (with its benefits and downsides) or to rebuild for every language.

**Reply by phitoduck** [reply]
Extreme idea: rewrite projen in a compiled language, e.g. Rust and then generate bindings in the target languages. Ship the bindings together with compiled binaries. 

Technically, that eliminates the need to have a 2nd language installed, and it's certainly improve the synth time haha.

Think of the marketing potential. Projen: blazing fast, typesafe, memory safe, requires you to be a rust genius to contribute to 🤣

**Reply by mrgrain** [reply]
I'd probably contribute just for fun. 😅

**Reply by phitoduck** [reply]
Man, I searched and searched. I can see why the JSII was created.

1. Besides the fact that more devs know JS than any other language,

3. It is *HARD* to set up a tool chain that can generate bindings in multiple languages for Golang or Rust ☹️. From what I could see, even if you could achieve it, you'd have to do lots of gymnastics to the Golang/Rust code which would make it hard both to read and write.

**Reply by mrgrain** [reply]
Yeah, honestly the one possible iteration of jsii I can see in the future is Web Assembly.

It would remove the need to install node. And it would likely possible to support development in different languages.

However what's not going to go away is the requirement for the jsii assembly (or some derivative of it). Because as you've discovered as well, jsii-pacmak is doing some heavy lifting here in generating consistent language bindings. So in order to support development in language X, we would need a compiler for language X to Wasm and to jsii assembly.

**Comment by mrgrain** [comment]
> 2.2 - projen new doesn't support projects contributed in Python
>
> We used cookiecutter to bootstrap our initial .projenrc.py and minimal Makefile that installed the Node/Python dependencies to run it.

It will highly depend on your goals, but I don't think this problem just goes away by moving off Node. The need for an inventory is arising because of a) the cli providing options as cli options, and b) the need to generate a .projenrc.py with preset options.

If you don't need either of these - great! You could probably build a very simple wrapper script around projen (or pygen) to do that. If you however need an inventory, you will find yourself in the need to build this for Python. This might be easy, or not. I honestly don't know. Projen uses jsii for it because it uses jsii anyway.

**Reply by phitoduck** [reply]
I guess if I'm arguing for a slimmer tool with more flexibility, then that likely puts me in the camp that doesn't find it a deal breaker if there is no "template inventory".

**Reply by mrgrain** [reply]
We'd definitely consider support for install from python directly. Due to lack of alternatives the inventory format would probably still be a jsii assembly, but you can easily fake an empty one. There are some other hurdles that aren't going to work, so someone will have to investigate that.

**Comment by mrgrain** [comment]
> 2.3 - Inferior escape hatching support
> We wanted all DE, MLOps, and MLE folks to be able/willing to contribute to our components. So we authored all our internal components in Python and vended them as a pip package.
> 
> The JSII doesn't expose private properties in the Python bindings.

I give you that in TS you can use private properties as hacky escape hatches. But in reality this just means that we are missing certain abstraction functionality in projen and I'd love to see the issues for it.

> Anecdote: We needed to patch arrays, e.g. in a GitHub Actions YamlFile, there is workflow.someJob.steps[]. We could never know for sure which index to modify. And we had no ability to inspect the content to find out.

This specifically has come up a bunch of times and projen definitely wants to support this in one way or another.

> This became reason the DS went from loving to resenting projen. ("If we use your tool, we're locked in to a setup that may not make sense for our project!")

Probably too late, but projen doesn't lock you in. You can always `projen eject` and worst case is you have used projen as a template engine.

**Reply by phitoduck** [reply]
> in reality this just means that we are missing certain abstraction functionality in projen and I'd love to see the issues for it

Yeah. I remember you and I chatting about this way back when. I'll think about it and maybe finally create that issue 😅.

I actually like that the `someObjectFile.obj` property is private. It could cause issues if people modified it willy-nilly. 

In the Python world, `thing._some_private_attr` communicates "if you insist on modifying this property, then fine, but you also assume responsibility to handle any breaking changes that we may release in the future upstream". 

I'd personally be down for that kind of implicit contract for modifying private `projen` attrs. That's sort of the whole philosophy behind escape hatching.

> You can always `projen eject`

Hah. I/we knew this. It was more that as the ML Platform lead, I was pressuring everyone to use our internal `projen` components 😅.

**Reply by mrgrain** [reply]
> I'd personally be down for that kind of implicit contract for modifying private projen attrs. That's sort of the whole philosophy behind escape hatching.

This is certainly a major downside of using jsii. In TS, we have this and it pretty much works exactly as you describe.

**Comment by mrgrain** [comment]
> 2.4 - Not great support for "multi-package mono repos"
> TL;DR - our project setup was so different from PythonProject, the only projen-specific features we used were escape hatching and the file management system in .projen/.

There's a reason why I'm hesitant to build more monorepo support directly into projen. If project setups are already opinionated, monorepo setups are even worse. We just won't get it right. To be blunt for a moment, even in your short list you already have two points that are purely opinionated:

> - was a lot to look at 😅
> - ci.yaml needed to include a "lint/format/test", section for each PythonPackage. Convenient to have these in one workflow rather than multiple.'

😉 

--- 

>   - we landed on 2 core abstractions: class Repo(Project) and class PythonPackage(Component)

I'm leaning towards a top-level `Repo` abstraction as well. Granted this is currently intermingled with projects. At the moment the parent project takes the place of the repo. It's great that you've found a way to solve this though!

> - .pre-commit-config.yaml needed to reflect information about all packages, e.g. "have black to ignore any files generated with projen across all Python packages"

Most tools in the Node world support hierarchical configurations, i.e. you would create the configs at the package level and they can potentially inherit a top-level config. The projen way would be that the pre-commit-config should not know about the packages. It should probably call a single `pre-commit-check` task in the top-level project. And that task should than take care of calling the `pre-commit-check` task for each subproject. Every suproject can than decide for itself what needs to be checked.

**Reply by phitoduck** [reply]
> in your short list you already have two points that are purely opinionated:

Fair! What I meant to get at was: we basically didn't get value from the existing opinionated `projen` components.

Thought:

1. most teams don't use off-the-shelf `cookiecutter` templates
2. they usually prefer to write their own from scratch--at most taking inspiration from publicly available templates

So maybe this means that there's a place for a lib in each language that only provides the `Component` abstraction.

**Reply by phitoduck** [reply]
I went and added a "Pros/Cons of Forking" section to the discussion head--inspired by this comment.

**Reply by mrgrain** [reply]
Irrespectively, I think your assessment is correct. Projen is currently two things: A framework that is used as the basis for other tools like [AWS PDK](https://github.com/aws/aws-pdk). But at the same time it is an opinionated batteries-included tool. I don't think this is much of a contradiction. But we will at some point certainly have to start thinking about how to "slice" separate these different roles a bit better.

**Comment by mrgrain** [comment]
> 2.5 No standard solution for escape hatching non-ObjectFiles

> Had to add the following components. Python users are familiar with the Jinja templating engine. For non-`ObjectFile`s like `Makefile`, `Dockerfile`,

Yeah that's fair. Both examples could probably be represented by a higher level abstraction or at least a list of instructions. In principle we are aware of the need to support hooks here. Maybe your solution (minus Jinja templates) could be a good starting point.

**Reply by phitoduck** [reply]
I saw `Makefile.ts` in `projen`. Respect for taking that on. I agree that's fundamentally more extensible. I'd also be down for a different templating language than Jinja if that were going to be a thing.

Downsides of Jinja-based components:
- It was hard to extend components built on templatized files. E.g. if you had `BentoMLDockerfile`, it was pretty much impossible inject more lines into the unrendered template. So if you wanted something like `ClassifierModelBentoMLDockerfile`... good luck. 
- It turns out BentoML generates Dockerfiles for you, and does so by exposing a Jinja template... so we had to use Jinja to generate Jinja. 

Positives:
- DS, MLE, MLOps, DE, and Analytics all knew it
- Syntax highlighting and autocompletion! Rather than edit a Dockerfile or Makefile through class methods, you were editing an actual Docker or Makefile, so the editor gave good highlighting, etc. Made editing files more accessible to these non-power users of projen.

**Reply by mrgrain** [reply]
Honestly I'm quite convinced that extending components can only ever work well on well-defined, but naturally limited integration points. Like a higher-level component wrapping a Dockerfile could expose a setting for software that is to be installed. Or of course it could define hooks (pre-install, post-install, pre-copy-files, post-copy-files). But these don't mean anything in a general sense. It's specific to the component.

The fact that escape hatches work for object files is imho a lucky side-effect that in object we can seantically address specific locations. A text file could do the something similar using byte or line/col positions. But who would want that.

**Comment by mrgrain** [comment]
> TL;DR The DevEx for projen in non-TS languages isn't as good. What would we lose if we had slimmer, Python-centric tool? Or could/should we modify projen?

I'm very sympathetic with the idea. I'm glad you found parts of projen useful, even if you didn't end up using the higher-level abstractions where for me most of the value is. Though it sounds a lot of it is just missing features in projen to meet your needs better.

I reckon you can easily build the slim downed version you described in a couple of days.

**Reply by tinovyatkin** [reply]
> TL;DR The DevEx for projen in non-TS languages isn't as good. What would we lose if we had slimmer, Python-centric tool? Or could/should we modify projen?

We’ve had great success using Projen in our Python-based CDK ML projects and monorepos. However, we don’t use any of the existing Python-specific Projen constructs, as they’re built around an outdated approach to Python development.

Projen really shines in CDK projects because CDK itself requires Node.js, and the build methodologies for CDK and Projen are very similar, allowing for knowledge reuse. We’ve also leveraged Projen constructs in our TypeScript (and polyglot) projects, where JSII helps avoid code duplication across languages.

That said, if your project doesn't use AWS CDK (or Terraform CDK) and is purely a Python project, you might miss out on some great Python-native tooling. For example, [Copier](https://copier.readthedocs.io/en/stable/updating/) offers updatable templates, and [PDM](https://pdm-project.org/latest/usage/template/) provides project templates. I also expect uv (the Rye successor) to eventually include template support, as their [Application projects](https://docs.astral.sh/uv/concepts/projects/#applications) concept seems like a step in that direction.

In any case, I agree that Projen (and JSII/CDK in general) could definitely be improved for Python developers. For Python-only or Python-specific projects, though, it might not be the ideal foundation.

**Reply by mrgrain** [reply]
I do think projen is definitely suffering from a lack of invested Python devs. The team maintaining it doesn't use Python regularly enough to have the required expertise to create a great Python experience.

**Reply by phitoduck** [reply]
@tinovyatkin Oh awesome, yeah.

No shade on `uv`, `pdm`, `copier`, `cookiecutter`, and `cdk init`. I think having a quick start to show you the right opinions is highly valuable.  Ultimately, I imagine a team would draw inspiration from those templates when creating their own library of `projen` components.

Oddly: we also used CDK, but not the projen template for it. We authored our CDK constructs in pure python and vended them to DS, MLE, DE, etc. that way. I felt that the CDK template was mainly applicable if you wanted to use the JSII/TS.

But we still found tremendous value in projen.

Basically, what `projen` has given our Python teams is:

1. A bulletproof upgrade path (I'd argue this is significantly more reliable, and easy to work with than copier because it takes a hard stance on avoiding "template drift")

2. A library of composable components with opinions that make sense for our team. E.g. when I want an additional AWS Lambda function in my project, I just add another `PythonPackage` and it's magically there, already set up with CI and any opinions used by my team.

Our DS and DE repos could be highly variable, so having components was a great way to impose constraints, while still being flexible and getting upgradability. 

I'll note: we had some early skeptics ("how often do we _reaaaally_ modify our templates?"), but the answer soon became "multiple times per week" because central project templates can be a great way to vend platform-level improvements all at once, ideally behind an abstraction.

**Reply by phitoduck** [reply]
> definitely suffering from a lack of invested Python devs @mrgrain 

Mmm. Yeah. I've liked some of the recent suggestions around modifying the existing Python-related components in projen, e.g. https://github.com/projen/projen/issues/3497

**Comment by phitoduck** [comment]
Followup: Thanks for the engagement @mrgrain and @tinovyatkin. Genuinely.

I'm still undecided. 

If an issue could get through to somehow provide Python with access to the `.obj` property of `ObjectFiles` (I know, weird that that's my main obstacle 🤣), then I may just continue using the `Component` engine in `projen` itself. 

I'll open that issue when I get time, and put some thought into what it might mean to improve the Python DevEx in projen without forking.
